### DISCUSSION TITLE
start async func with arguments without calling func on init

### DISCUSSION THREAD

**Discussion by HilkopterBob** [zauberzeug/nicegui#673] [Q&A] [answered]
hey, i want to call async functions with arguments without starting them on init.

situation:
i want three checkboxes from wich only one can be checked at all times or none. If none is selected, a notification should pop up.
code:

```python
def hosts_changed_values(changed_component, second_component, third_component):
    if changed_component.value != False:
        second_component.set_value(not changed_component.value)
        third_component.set_value(not changed_component.value)
    if changed_component.value == False and second_component.value == False and third_component.value == False:
        sleep(5)
        ui.notify("no host-type selected", type="warning")

host_all = ui.checkbox("", value=True).on("click", lambda: hosts_changed_values(host_all, host_groups, host_individual))
host_groups = ui.checkbox("", value=False).on("click", lambda: hosts_changed_values(host_groups, host_all, host_individual))
host_individual = ui.checkbox("", value=False).on("click", lambda: hosts_changed_values(host_individual, host_all, host_groups))
```

this workes great, besides that the last deactivated checkbox lags for the sleep period.

so i made the function async:
```python
async def hosts_changed_values(changed_component, second_component, third_component):
    if changed_component.value != False:
        second_component.set_value(not changed_component.value)
        third_component.set_value(not changed_component.value)
    if changed_component.value == False and second_component.value == False and third_component.value == False:
        await asyncio.sleep(5)
        ui.notify("no host-type selected", type="warning")
```

but now whenever i call the method i get following error:
RuntimeWarning: coroutine 'hosts_changed_values' was never awaited
  if is_coroutine(handler):
RuntimeWarning: Enable tracemalloc to get the object allocation traceback

now the problem:

its not possible to call async funcs with lambda as they will never be awaited
ibut i need the lambda so the func doesnt get executed on init. The workaround would be to strap back the parentheses and kwargs
but the function needs them to properly work.
So is there a way to work around that?

**Comment by falkoschindler** [comment, marked as answer]
That's a common problem with `lambda` expressions in Python: You can't make them async.

One way around it is to define a wrapper function (like `sleep3` in the example below) that expects no arguments:

```py
async def sleep(duration: float):
    ui.notify('Sleeping...')
    await asyncio.sleep(duration)
    ui.notify('Done.')

async def sleep3():
    await sleep(3)

ui.button('Sleep 3 seconds', on_click=sleep3)
```

An alternative is to use `partial` from the `functools` package:

```py
from functools import partial

async def sleep(duration: float):
    ui.notify('Sleeping...')
    await asyncio.sleep(duration)
    ui.notify('Done.')

ui.button('Sleep 3 seconds', on_click=partial(sleep, 3))
```

**Reply by HilkopterBob** [reply]
hey worked like a charm! 
besides that, why are you moving issues to discussions?

**Reply by falkoschindler** [reply]
We're trying to separate Q&A and other discussions from actual issues with NiceGUI. So the community can find questions like yours more easily, can comment and vote on answers and see which answer you accepted (which is not possible with issues). Meanwhile we can focus on what needs to get fixed, assign developers and milestones and so on (which is not possible with discussions).
Momentarily I'm working on issue templates to inform users about the distinction upfront.
